Identity center resource sync event definition and UI formatter

flyinghermit opened the pull request:
This PR adds audit event proto type and UI formatter for Identity Center resource sync event. The resource sync runs periodically, every 5 minutes.

code: `TAIC001I`
event": `aws_identity_center.resource_sync.success`
code: `TAIC001E`
event": `aws_identity_center.resource_sync.failed`

A sample event message:
```
{
  "code": "TAIC001I",
  "event": "aws_identity_center.resource_sync",
  "message": "Account, Permission Set, Account Assignments imported",
  "success": true,
  "time": "2025-02-11T15:29:30.045Z",
  "total_account_assignments": 12,
  "total_accounts": 4,
  "total_permission_sets": 3,
  "total_user_groups": 3,

}
```



Part of https://github.com/gravitational/teleport.e/issues/5513

flyinghermit committed:
feat: identity center resource sync event

flyinghermit committed:
Web UI event formatter

flyinghermit committed:
let the emitter send user message to be used in the ui

flyinghermit commented:
Only total imported resource counts are included in the audit event. The initial plan was to include resource id and ARN but I realize that may account for too much data for event that runs every 5 minutes. Say 500 account, 10 permission and 5000 account assignments. That is already a heavy message as an event item. We can include them later if our users really want to see them.

flyinghermit committed:
Merge branch 'master' of github.com:gravitational/teleport into sshah/awsic-sync-event

flyinghermit commented:
changelog will be added in the backport as I will backport all identity center related events at once.

smallinsky commented on api/proto/teleport/legacy/types/events/events.proto:
```diff
@@ -8081,3 +8082,29 @@ message StableUNIXUser {
+    (gogoproto.jsontag) = ""
+  ];
+  // Total imported AWS accounts.
+  int32 total_account = 2 [(gogoproto.jsontag) = "total_account,omitempty"];
```
```suggestion
  int32 total_accounts = 2 [(gogoproto.jsontag) = "total_account,omitempty"];
```

Can we use plural ? that will make the fields naming consitent with other UsageEvents like: 
https://github.com/gravitational/teleport/blob/7b6f291fb348850f7e621a5e102fa0094f32bdd1/api/proto/teleport/usageevents/v1/usageevents.proto#L753

flyinghermit replied:
https://github.com/gravitational/teleport/pull/52059/commits/980ba2a98b9dc19b87d962caa6490ee39101b768

smallinsky commented on api/proto/teleport/legacy/types/events/events.proto:
```diff
@@ -8081,3 +8082,29 @@ message StableUNIXUser {
+  // Total permission sets imported from Identity Center.
+  int32 total_permission_set = 5 [(gogoproto.jsontag) = "total_permission_set,omitempty"];
+  // Status is a resource sync status.
+  Status Status = 6 [
+    (gogoproto.nullable) = false,
+    (gogoproto.embed) = true,
+    (gogoproto.jsontag) = ""
+  ];
```
Will we have an equivalent of:
```
// OktaSyncFailureEvent is emitted when the Okta synchronization fails.
	OktaSyncFailureEvent = "okta.sync.failure"
```	
where the failure event has a different FE code prefix do display failed even with Warning Icon ?

The Status Status is only validated when the sync fails.

flyinghermit replied:
Implemented dedicated failure event https://github.com/gravitational/teleport/pull/52059/commits/bd1104570c3086009bc96b717995426fcb342430

flyinghermit committed:
review: plural total count names

flyinghermit committed:
review: make distinctive failure event

smallinsky commented on web/packages/teleport/src/services/audit/makeEvent.ts:
```diff
@@ -2029,6 +2029,20 @@ export const formatters: Formatters = {
   },
+  [eventCodes.AWS_IC_RESOURCE_SYNC_SUCCESS]: {
+    type: 'aws_identity_center.resource_sync.success',
+    desc: 'AWS IAM Identity Center Resource Sync',
```
nit: 
```suggestion
    desc: 'AWS IAM Identity Center Resource Sync Completed',
```

flyinghermit replied:
Done https://github.com/gravitational/teleport/pull/52059/commits/53881d50a53f6e737578b26633feb193b9fcc27c#

smallinsky commented on web/packages/teleport/src/services/audit/makeEvent.ts:
```diff
@@ -2029,6 +2029,20 @@ export const formatters: Formatters = {
+  },
+  [eventCodes.AWS_IC_RESOURCE_SYNC_FAILURE]: {
+    type: 'aws_identity_center.resource_sync.failed',
+    desc: 'AWS IAM Identity Center Resource Sync',
```
nit:
```suggestion
    desc: 'AWS IAM Identity Center Resource Sync Failed',
```

smallinsky commented on lib/events/codes.go:
```diff
@@ -705,6 +705,13 @@ const (
+	AWSICResourceSyncSuccessCode = "TAIC001I"
+	// AWSICResourceSyncFailureCode is the AWS Identity Center resource
+	// sync attempt failure code.
+	AWSICResourceSyncFailureCode = "TAIC002I"
```
```suggestion
	AWSICResourceSyncFailureCode = "TAIC002E"
```

The FE has some login that distinguish event type severity the the  event code suffix  where

W - stand for warning severity
I - Info
E - Error

flyinghermit replied:
yeah my bad. I missed that when I copy pasted field to resolve https://github.com/gravitational/teleport/pull/52059#discussion_r1953247064. 
See https://github.com/gravitational/teleport/pull/52059/commits/53881d50a53f6e737578b26633feb193b9fcc27c#diff-763355b55b25168aa0b951fdcba75437138edc2a804c19d90a9b74d80dfe7c96R713

smallinsky commented on web/packages/teleport/src/services/audit/makeEvent.ts:
```diff
@@ -2029,6 +2029,20 @@ export const formatters: Formatters = {
+    type: 'aws_identity_center.resource_sync.success',
+    desc: 'AWS IAM Identity Center Resource Sync',
+    format: ({ message }) => {
+      return message;
```
nit: 

Printing message will make the UI vague: 

![Screenshot 2025-02-12 at 20 45 36](https://github.com/user-attachments/assets/cf928f81-d51a-4688-ad79-e804c52099bc)


The User Audit Event description should be meaningful to the user. I suggest extracting relevant data from the audit event and displaying something like:

AWS IC synchronization successfully completed [groups: X, permission sets: X, account: X]

This way, the user can immediately understand the details of the audit event.

flyinghermit replied:
Done https://github.com/gravitational/teleport/pull/52059/commits/53881d50a53f6e737578b26633feb193b9fcc27c#diff-5018bbae0e65b4c5bddd40edf2b82046daeac34a2872f97b9c4e0d58d183681fR2045

smallinsky reviewed (commented):
LGTM when the nit comment will be addressed.

flyinghermit committed:
review suggestion:
- spell out completed and failed event description
- write resource import counts
- use  code for error

gzdunek commented on api/proto/teleport/legacy/types/events/events.proto:
```diff
@@ -8081,3 +8082,29 @@ message StableUNIXUser {
+  // Total imported AWS accounts.
+  int32 total_accounts = 2 [(gogoproto.jsontag) = "total_accounts,omitempty"];
+  // Total account assignments imported from Identity Center. Account assignments
+  // are permission's assigned to users or group.
```
```suggestion
  // are permissions assigned to users or group.
```

flyinghermit replied:
Done https://github.com/gravitational/teleport/pull/52059/commits/cb2e67cc1ee42b1201986a8ebcdd0c5aa72fcf22

flyinghermit committed:
spell fix

flyinghermit committed:
Merge branch 'master' into sshah/awsic-sync-event
